src/main/scala/corpus.scala

```diff
@@ -0,0 +1,87 @@
+package treesitter.lsp
+
+import parsley.character.*
+import parsley.debug.*
+import parsley.combinator.*
+import parsley.Parsley
+
+enum LispNode:
+  case Leaf(name: String)
+  case Nest(title: String, nodes: List[LispNode])
+
+object LispNode:
+  export parsers.parser
+
+  private object parsers:
+    def constructNode(nm: String, subnodes: List[LispNode]) =
+      subnodes match
+        case head :: next => LispNode.Nest(nm, subnodes)
+        case Nil          => LispNode.Leaf(nm)
+
+    val nodeName = some(letter <|> char('_')).map(_.mkString)
+
+    extension [A](p: => Parsley[A]) def wsp = p <* whitespaces
+
+    lazy val parser: parsley.Parsley[LispNode] = (
+      char('(') *>
+        (nodeName.wsp <~> {
+
+          val rec = many(parser)
+
+          val simple = nodeName
```

Comment:
A `SingleLeaf` parser bridge could make this `SingleLeaf(nodeName)`, which is cute.